spec/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -599,41 +604,77 @@ function recvPacket(
 
     // all assertions passed (except sequence check), we can alter state
 
-    // always set the acknowledgement so that it can be verified on the other side
-    provableStore.set(
-      packetAcknowledgementPath(packet.destPort, packet.destChannel, packet.sequence),
-      hash(acknowledgement)
-    )
-
     if (channel.order === ORDERED) {
       nextSequenceRecv = provableStore.get(nextSequenceRecvPath(packet.destPort, packet.destChannel))
       abortTransactionUnless(packet.sequence === nextSequenceRecv)
       nextSequenceRecv = nextSequenceRecv + 1
       provableStore.set(nextSequenceRecvPath(packet.destPort, packet.destChannel), nextSequenceRecv)
-    } else
-      abortTransactionUnless(provableStore.get(packetAcknowledgementPath(packet.destPort, packet.destChannel, packet.sequence) === null))
+    } else {
+      // for unordered channels we must set the receipt so it can be verified on the other side
+      // this receipt does not contain any data, since the packet has not yet been processed
+      // it's just a single store key set to an empty string to indicate that the packet has been received
+      abortTransactionUnless(provableStore.get(packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence) === null))
+      provableStore.set(
+        packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence),
+        ""
+      )
+    }
 
-    // log that a packet has been received & acknowledged
-    emitLogEntry("recvPacket", {sequence: packet.sequence, timeoutHeight: packet.timeoutHeight,
-                                timeoutTimestamp: packet.timeoutTimestamp, data: packet.data, acknowledgement})
+    // log that a packet has been received
+    emitLogEntry("recvPacket", {sequence: packet.sequence, timeoutHeight: packet.timeoutHeight, port: packet.destPort, channel: packet.destChannel,
+                                timeoutTimestamp: packet.timeoutTimestamp, data: packet.data})
 
     // return transparent packet
     return packet
 }
 ```
 
-#### Acknowledgements
+#### Writing acknowledgements
+
+The `writeAcknowledgement` function is called by a module in order to write data which resulted from processing an IBC packet that the sending chain can then verify, a sort of "execution receipt" or "RPC call response".
+
+Calling modules MUST execute application logic atomically in conjunction with calling `writeAcknowledgement`.
+
+This is an asynchronous acknowledgement, the contents of which do not need to be determined when the packet is received, only when processing is complete. In the synchronous case, `writeAcknowledgement` can be called in the same transaction (atomically) with `recvPacket`.
```

Comment:
Okay, so in the "normal" case we have today, we just call `writeAcknowledgement` in the OnRecv handler rather than return the ack bytes, correct? 

This may cause hard to debug errors for people's code if they don't follow this issue. It would be good to maintain a simple migration guide for all IBC changes that affect people building on IBC. (I asked for a migration guide to the sdk as well. At least IBC is more or less stable since some weeks, and makes sense to keep a diff of what changed since some recent point in time)

Reply:
That's correct; to replicate the current behaviour simply call `writeAcknowledgement` synchronously.